src/ray/core_worker/task_state_buffer.cc

```diff
@@ -0,0 +1,175 @@
+                                       1000,  // TODO(rickyx) use a flag
+                                       "CoreWorker.deadline_timer.flush_task_events");
+}
+
+void TaskStateBuffer::AddTaskEvent(TaskID task_id,
+                                   const TaskSpecification &task_spec,
+                                   rpc::TaskStatus task_status) {
+  absl::MutexLock lock(&mutex_);
+
+  auto task_events_itr = AddTaskEventInternal(task_id, task_status);
+  RAY_CHECK(task_events_itr != task_events_map_.end())
+      << "Failed to initialize a new task events";
+
+  if (ShouldIncludeTaskInfo(task_status)) {
+    // Fill the task info for task metadata
+    RAY_CHECK(task_spec.TaskId() == task_id) << "Unmatched task ids";
+    auto task_info = FillTaskInfo(task_spec);
+    task_events_itr->second.mutable_task_info()->Swap(&task_info);
+  }
+}
+
+void TaskStateBuffer::AddTaskEvent(TaskID task_id, rpc::TaskStatus task_status) {
+  absl::MutexLock lock(&mutex_);
+
+  auto task_events_itr = AddTaskEventInternal(task_id, task_status);
+  RAY_CHECK(task_events_itr != task_events_map_.end())
+      << "Failed to initialize a new task events";
+}
+
+TaskIdEventMap::iterator TaskStateBuffer::AddTaskEventInternal(
+    TaskID task_id, rpc::TaskStatus task_status) {
+  auto task_events_itr = GetOrInitTaskEvents(task_id, task_status);
+
+  auto event = task_events_itr->second.add_events();
+  event->set_task_status(task_status);
+  event->set_event_time(absl::GetCurrentTimeNanos());
+
+  return task_events_itr;
+}
+
+rpc::TaskInfoEntry TaskStateBuffer::FillTaskInfo(const TaskSpecification &task_spec) {
+  rpc::TaskInfoEntry task_info;
+  rpc::TaskType type;
+
+  if (task_spec.IsNormalTask()) {
+    type = rpc::TaskType::NORMAL_TASK;
+  } else if (task_spec.IsActorCreationTask()) {
+    type = rpc::TaskType::ACTOR_CREATION_TASK;
+    task_info.set_actor_id(task_spec.ActorCreationId().Binary());
+  } else {
+    RAY_CHECK(task_spec.IsActorTask());
+    type = rpc::TaskType::ACTOR_TASK;
+    task_info.set_actor_id(task_spec.ActorId().Binary());
+  }
+  task_info.set_type(type);
+  task_info.set_name(task_spec.GetName());
+  task_info.set_language(task_spec.GetLanguage());
+  task_info.set_func_or_class_name(task_spec.FunctionDescriptor()->CallString());
+  // TODO(rickyx): we will have scheduling states recorded in the events list.
+  task_info.set_scheduling_state(rpc::TaskStatus::NIL);
+  task_info.set_job_id(task_spec.JobId().Binary());
+
+  task_info.set_task_id(task_spec.TaskId().Binary());
+  task_info.set_parent_task_id(task_spec.ParentTaskId().Binary());
+  const auto &resources_map = task_spec.GetRequiredResources().GetResourceMap();
+  task_info.mutable_required_resources()->insert(resources_map.begin(),
+                                                 resources_map.end());
+  task_info.mutable_runtime_env_info()->CopyFrom(task_spec.RuntimeEnvInfo());
+
+  // TODO(rickyx): Add node id later
+  // if (!node_id.IsNil()) {
```

Comment:
I guess we should have metadata & mutable data separation and this can be updated as a mutable field

Reply:
but i think in terms of performance, there's minimal impact?